File: leetcode/leetcode46.java
// idea: use visited to judge that whether current number has been added to list
// we don't use start index here because each time of recursion we would iterate nums again
// beats over 91% of submission
class Solution {
    private void dfs(int[] nums, boolean[] visited, List<Integer> sublist, List<List<Integer>> result){
        if(sublist.size() == nums.length){
            result.add(new ArrayList<>(sublist));
            return;
        }
        
        for(int i = 0; i < nums.length; i++){
            if(visited[i]) continue;
            sublist.add(nums[i]);
            visited[i] = true;
            dfs(nums, visited, sublist, result);
            sublist.remove(sublist.size()-1);
            visited[i] = false;
        }
    }
    
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        if(nums == null) return result;
        
        dfs(nums, new boolean[nums.length], new ArrayList<>(), result);
        
        return result;
    }
}

File: leetcode/leetcode46.cpp
// idea: use visited to judge that whether current number has been added to list
// we don't use start index here because each time of recursion we would iterate nums again
// beats over 16% of submission
class Solution {
    private:
        void dfs(vector<int>& nums, vector<bool>& visited, vector<int>& sub, vector<vector<int>>& res){
            if(sub.size() == nums.size()){
                res.push_back(sub);
                return;
            }
            
            for(int i = 0; i < nums.size(); ++i){
                if(visited.at(i)) continue;
                sub.push_back(nums.at(i));
                visited.at(i) = true;
                dfs(nums, visited, sub, res);
                sub.pop_back();
                visited.at(i) = false;
            }
        }
    public:
        vector<vector<int>> permute(vector<int>& nums) {
            vector<int> sub;
            vector<bool> visited(nums.size(), false);
            vector<vector<int>> res;
            
            dfs(nums, visited, sub, res);
            
            return res;
        }
    };
}

// idea: https://www.geeksforgeeks.org/write-a-c-program-to-print-all-permutations-of-a-given-string/
// beats over 86% of submission
class Solution {
private:
    void dfs(vector<int>& nums, int start, vector<vector<int>>& res){
        if(start == nums.size() - 1){
            res.push_back(nums);
            return;
        }
        
        for(int i = start; i < nums.size(); ++i){
            swap(nums.at(start), nums.at(i));
            dfs(nums, start+1, res);
            swap(nums.at(start), nums.at(i));
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        
        dfs(nums, 0, res);
        
        return res;
    }
};